### PR TITLE
Add dfs-collect :pre/:post modes, custom reduce, and comprehensive tests

### PR DESCRIPTION
**Pull request by H31MDALLR**
## Summary

- Extend `dfs-collect` with `:pre`/`:post` collection modes and optional `reduce-fn`/`init` parameters
- Add comprehensive tests for `dfs-collect`, `dfs-find`, and edge cases
- Extract duplicate callback logic into shared `maybe-collect` helper

## Changes

**`graph.clj`:**
- Add `collect-on-visit?` predicate helper
- Support `:pre` (alias for `:visit`) and `:post` (children-before-parents) collection modes
- Accept optional `reduce-fn` and `init` parameters for custom accumulation
- Extract triplicated on-visit/on-cycle/on-missing callbacks into `maybe-collect`

**`graph_test.clj` / `interface_test.clj`:**
- Tests for pre-order and post-order collection
- Tests for custom reduce functions (counting, set accumulation)
- Edge case coverage: empty graphs, self-loops, disconnected components, deep graphs

## Test plan

- [ ] `bb test` passes
- [ ] Pre-existing `dfs` and `dfs-validate-graph` tests unaffected
- [ ] New `dfs-collect` tests cover all `:collect-on` modes

🤖 Generated by Miniforge SDLC workflow (first autonomous PR!)

### CODE REVIEW

File: components/algorithms/src/ai/miniforge/algorithms/graph.clj [hunk 1]
```diff
@@ -180,51 +180,111 @@
...
+                 #{}
+                 start-ids))
+       @acc)
+     ;; Pre-order and other modes: collect during traversal events.
+     (let [collected (atom init)]
+       (dfs graph
+            start-ids
+            get-deps-fn
+            ;; on-visit
+            (fn [node-id _node path visited visiting]
```
Comment (H31MDALLR) [inline components/algorithms/src/ai/miniforge/algorithms/graph.clj:274]: These functions here look somewhat similar, except for the predicates. Yeah, pull this out to a function and pass in the predicate.

Comment (H31MDALLR) [reply components/algorithms/src/ai/miniforge/algorithms/graph.clj:274]: Fixed in db8a906 — extracted `maybe-collect` closure that takes the `active?` predicate. The three callbacks now delegate to it instead of duplicating the collect-fn/swap!/nil pattern.
